monai/optimizers/utils.py

```diff
@@ -48,27 +50,31 @@ def generate_param_groups(
             match_types=["select", "filter"],
             lr_values=[1e-2, 1e-3],
         )
+        # the groups will be a list of dictionaries:
+        # [{'params': <generator object Module.parameters at 0x7f9090a70bf8>, 'lr': 0.01},
+        #  {'params': <filter object at 0x7f9088fd0dd8>, 'lr': 0.001},
+        #  {'params': <filter object at 0x7f9088fd0da0>}]
         optimizer = torch.optim.Adam(params, 1e-4)
 
     """
     layer_matches = ensure_tuple(layer_matches)
-    match_types = ensure_tuple(match_types)
-    lr_values = ensure_tuple(lr_values)
-    if len(layer_matches) != len(lr_values) or len(layer_matches) != len(match_types):
-        raise ValueError("length of layer_match callable functions, match types and LR values should be the same.")
+    match_types = ensure_tuple_rep(match_types, len(layer_matches))
+    lr_values = ensure_tuple_rep(lr_values, len(layer_matches))
 
     params = list()
     _layers = list()
     for func, ty, lr in zip(layer_matches, match_types, lr_values):
-        if ty == "select":
+        if ty.lower() == "select":
             layer_params = func(network).parameters()
-        elif ty == "filter":
+            layer_ids = func(network).parameters()
+        elif ty.lower() == "filter":
             layer_params = filter(func, network.named_parameters())
+            layer_ids = filter(func, network.named_parameters())
         else:
-            raise ValueError(f"unsuppoted layer match type: {ty}.")
+            raise ValueError(f"unsupported layer match type: {ty}.")
 
         params.append({"params": layer_params, "lr": lr})
-        _layers.extend(list(map(id, layer_params)))
+        _layers.extend(list(map(id, layer_ids)))
```

Comment:
Can we change to `layer_params.deepcopy()` to avoid duplicated `func` and `filter` calls?

Thanks.

Reply:
thanks, I added local functions to return generators, that would work like making fresh copies.